### PR TITLE
222 wires

### LINKED ISSUE #222: Wires
### AC
- A wire takes up on block
  - Wires will connect if they are in adjacent blocks to each other.
  - A wire will be rendered & collide differently depending on its connections
    - [x] Create framework for rendering connected textures
    - [x] Create framework for rendering connected models
    - [x] Create framework for having connected colliders
  - With any connections, it will render as connection between them

### PR DESCRIPTION
**Pull request by DanielStebbins**
Completed:
- Logic graph with wires, logic on block, and logic indicator light.
- Logic groups that merge and split as wires connect them.
- Logic events for outputs and inputs (in that order) in each frame.

To do (issues created):
- Logic gates.
- Colored logic wires.
- Logic interactions with existing blocks, for example, missile launchers.

### CODE REVIEW

File: cosmos_client/assets/cosmos/blocks/logic_indicator.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+    "texture": {
+        "Sides": {
```
Comment (AnthonyTornetta) [inline cosmos_client/assets/cosmos/blocks/logic_indicator.json]: Can be "Single": "cosmos:light" I think

File: cosmos_core/src/block/specific_blocks/logic_indicator.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    fn indicator_on(&self) -> bool;
+    /// Turns the light on.
+    fn set_indicator_on(&mut self);
+    /// Turns the light off.
+    fn set_indicator_off(&mut self);
+}
+
+impl LogicIndicatorBlockInfo for BlockInfo {
+    fn indicator_on(&self) -> bool {
+        (self.0 >> 7) & 1 == 0
```
Comment (AnthonyTornetta) [inline cosmos_core/src/block/specific_blocks/logic_indicator.rs]: Put this "7" into a constant so it's not a magic number

File: cosmos_core/src/logic/logic_graph.rs [hunk 3]
```diff
@@ -0,0 +1,797 @@
...
+    pub local_face: BlockFace,
+}
+
+impl Port {
+    /// Convenience constructor for Ports.
+    pub fn new(coords: BlockCoordinate, local_face: BlockFace) -> Port {
+        Port { coords, local_face }
+    }
+
+    fn all_for(coords: BlockCoordinate) -> HashSet<Port> {
```
Comment (AnthonyTornetta) [inline cosmos_core/src/logic/logic_graph.rs]: Use HashSet::from_iter here

File: cosmos_core/src/logic/logic_graph.rs [hunk 4]
```diff
@@ -0,0 +1,797 @@
...
+        port_type: PortType,
+        structure: &Structure,
+        blocks: &Registry<Block>,
+        logic_blocks: &Registry<LogicBlock>,
+        evw_logic_input: &mut EventWriter<LogicInputEvent>,
+    ) {
+        let local_face = structure.block_rotation(coords).global_to_local(face);
+
+        // If the neighbor coordinates don't exist, no port is removed.
+        if let Ok(neighbor_coords) = coords.step(local_face) {
```
Comment (AnthonyTornetta) [inline cosmos_core/src/logic/logic_graph.rs]: Swap to `let else` for less indentation in method

File: cosmos_core/src/logic/logic_graph.rs [hunk 5]
```diff
@@ -0,0 +1,797 @@
...
+                structure,
+                blocks,
+                logic_blocks,
+                evw_logic_output,
+                evw_logic_input,
+            )
+        }
+
+        // Connect wire faces to all existing groups (by creating one new group that includes all adjacent groups).
+        if logic_block.wire_faces().count() > 0 {
```
Comment (AnthonyTornetta) [inline cosmos_core/src/logic/logic_graph.rs]: does this have the .is_empty() method? also early return here would be ideal for less indentation

File: cosmos_core/src/logic/logic_graph.rs [hunk 6]
```diff
@@ -0,0 +1,797 @@
...
+                structure,
+                blocks,
+                logic_blocks,
+                evw_logic_input,
+            )
+        }
+
+        // For wire faces, 1 connection means just delete the wire. 2+ means delete the wire's group and make a new one for each connection.
+        // For now, we just delete the group and start again every time to avoid edge cases.
+        if logic_block.wire_faces().count() > 0 {
```
Comment (AnthonyTornetta) [inline cosmos_core/src/logic/logic_graph.rs]: same as above

File: cosmos_core/src/logic/mod.rs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    blocks: Res<Registry<Block>>,
+    logic_blocks: Res<Registry<LogicBlock>>,
+    mut q_wire_graph: Query<&mut LogicGraph>,
+    mut q_structure: Query<&mut Structure>,
+    mut evw_logic_output: EventWriter<LogicOutputEvent>,
+    mut evw_logic_input: EventWriter<LogicInputEvent>,
+) {
+    for ev in evr_block_updated.read() {
+        // If was logic block, remove from graph.
+        if let Some(logic_block) = logic_blocks.from_id(blocks.from_numeric_id(ev.old_block).unlocalized_name()) {
```
Comment (AnthonyTornetta) [inline cosmos_core/src/logic/mod.rs:181]: let else for less indent.

File: cosmos_core/src/structure/coordinates.rs [hunk 8]
```diff
@@ -187,7 +212,7 @@ macro_rules! create_coordinate {
         #[doc=$structComment]
```
Comment (AnthonyTornetta) [inline cosmos_core/src/structure/coordinates.rs:212]: Please derive default for this too then

**Review by AnthonyTornetta — CHANGES_REQUESTED** [review]

**Comment by AnthonyTornetta** [comment]
Lot of expects in logic_graph.rs, at some point we may want to go back and see if we can rearchitect the graph a bit to remove some of them, and not represent as many invalid states.  But for a first creation of a logical/wire system, this is great.

File: cosmos_core/src/logic/mod.rs (hunk 7 above)
Comment (DanielStebbins) [reply cosmos_core/src/logic/mod.rs:181]: No, I like the indent

File: cosmos_client/src/ui/components/text_input.rs [hunk 9]
```diff
@@ -161,9 +161,9 @@ fn monitor_clicked(
 
             if *interaction == Interaction::Pressed {
                 focused.0 = Some(ent);
-            } else if focused.0.map(|x| x == ent).unwrap_or(false) {
+                // } else if focused.0.map(|x| x == ent).unwrap_or(false) {
                 // Only clear out focused ent if it was one of the ones we're worried about
-                focused.0 = None;
+                // focused.0 = None;
```
Comment (AnthonyTornetta) [inline cosmos_client/src/ui/components/text_input.rs:166]: Uncomment this

**Review by AnthonyTornetta — CHANGES_REQUESTED** [review]

**Review by AnthonyTornetta — APPROVED** [review]

**Pull request merged**